### ISSUE TITLE
[no-unused-vars] False positive when using Declaration Merging

### ISSUE THREAD

**Issue by n8eloy** [typescript-eslint/typescript-eslint#3573]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error",
  }
}
```

```TS
interface MyInterface<T> {
  [k: string]: T
}

export interface MyMergedInterface<T> extends MyInterface<T> {}

export class MyMergedInterface<T> {} // 'T' is defined but never used.
```

**Expected Result**

I did not expect any errors to be shown since the `T` variable is being used by `MyMergedInterface` first declaration.

**Actual Result**

Output is as follows for line 7 as commented in code above:

`'T' is defined but never used. eslint(@typescript-eslint/no-unused-vars)`

**Additional Info**

Hey all! I'm honestly unsure whether this issue is a false positive or just expected behavior. We are currently simply ignoring the warning through `eslint-disable-next-line`.

It's important to note that TS expects both merged declarations to have identical type parameters so we've found no "clean" workarounds here else than silencing the rule.

We have found [this comment](https://github.com/typescript-eslint/typescript-eslint/issues/2523#issuecomment-691710593) on a similar issue, but removing `<T>` from the second declaration doesn't seem to be valid and leads to an error, which is weird since we would have to choose between "ignoring" a TS error or ignoring a linter error for fully using this rule.

`All declarations of 'MyMergedInterface' must have identical type parameters. (ts(2428))`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.0` |
| `@typescript-eslint/parser`        | `4.28.0` |
| `TypeScript`                       | `4.3.4` |
| `ESLint`                           | `7.24.0` |
| `node`                             | `14.16.0` |

**Comment by bradzacher** [comment]
Worth mentioning that declaration merging interfaces into classes is [VERY unsafe and thus should be avoided at all costs](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgGIHt3IN7IA5Tp4BcyAzmFKAObIC+AUAgDZxllqY4PLLpgALaAAVCJcpRoBuZAHpZyQKDkyACoBlZNEJRkAWmSii0MAE9kAcn5CohvOeQC2yEFlDAwwOM2AAvaMjgQABNkYA4XMGQgiBg3SGYzNjJgahAIENBkQRQEdBAKKABXBDB0KAA6HjkFQBlyAzFjM3MCIntHcKx1TShtBkYGeVUBMORc6OQAd2BmZmcIADd-R0WA1Q0tMoAaAODQyKmZucWdACMUCjgPBC8E5GPgWLgT27I4GAgBhQnhhAFQjjg+EIzwgAFs9AEOO5zBwIGwzKVyBAUIJLpDJlQwJAQFlhhwDrNBIQJgFIkVwMBQRBKrl8pEAB7IAC8cxJGHQAAoAJTlFp4crMCAgaiCKQMIA).

----

As mentioned in the comment you linked - there's nothing we can do about this, unfortunately.
We can only determine if variables are used statically in a single file.
We can also only analyse statically within the current or nested scopes.

`T` is declared in the class's declaration scope.
`T` is not used within the class's declaration scope or any nested scopes.
Thus `T` in the class is unused.

As per the other comment - if you rely on this pattern and in fact need the variable for some reason - use a disable comment and explain why it's necessary.

**Comment by n8eloy** [comment]
> Worth mentioning that declaration merging interfaces into classes is VERY unsafe and thus should be avoided at all costs.

Unfortunately we had to do such in order to deal with an external library without polluting our own code base. The risks are mitigated by the way such lib work, but thanks for the heads-up

___

Seems like disabling the rule is the way to go then. Thank you for the quick response!

**Comment by AlansCodeLog** [comment]
Was just about to file the same issue. Too bad it has no solution. Although some note on the rule would be useful.

As a workaround so as to not litter my code with ignores and explanations, I created a type:

```ts
/**
 * @internal
 * See https://github.com/typescript-eslint/typescript-eslint/issues/3573
 */
// @ts-expect-error .
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type IgnoreUnusedTypes<T = any[]> = {}
```

Then in any class or interface with unused types:
```ts
class SomeClass<T> extends IgnoreUnusedTypes<[T]> {}
// or
interface SomeClass<T> extends IgnoreUnusedTypes<[T]> {}
```
